### PR TITLE
feat: validating configs before trying to intialize sdk's

### PR DESCRIPTION
**Pull request by anantjain45823**
## PR Description

Our motive is to validate configs before initialisation for all the integrations. 

## Notion ticket

Ticket link

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: src/core/analytics.js [hunk 1]
```diff
@@ -342,6 +354,7 @@ class Analytics {
               // logger.debug(msg);
               this.errorReporting.leaveBreadcrumb(msg);
               intgInstance = new intMod[modName](intg.config, self, intg.destinationInfo);
+              this.validateConfig(intgInstance.getRequiredConfigs(), intg.config, intg.name);
```
Comment (saikumarrs) [inline src/core/analytics.js:357]: `getRequiredConfigs` is not an appropriate term. `getRequiredConfigFields` is probably what you meant to use.

Comment (saikumarrs) [inline src/core/analytics.js:357]: Not all destinations have implemented the `getRequiredConfigs` method. So, initialization will fail for them right way.
We should first check if that method is defined.

Comment (saikumarrs) [reply src/core/analytics.js:357]: Moreover, it'll also not work if the user has locked the integrations version.

File: src/core/analytics.js [hunk 2]
```diff
@@ -195,6 +196,17 @@ class Analytics {
...
   }
 
+  validateConfig(requiredConfigsList, configsFromDashboard, intg) {
+    if(!requiredConfigsList){
+      throw Error(`${intg}: required configs not present`);
+    }
+    requiredConfigsList.forEach(config => {
+      const configVal = configsFromDashboard[config];
+      if (!isDefinedAndNotNull(configVal) || (typeof configVal === "string" && configVal.length === 0))
+        throw Error(`${intg}: ${config} is either not present or does not have valid value`);
```
Comment (saikumarrs) [inline src/core/analytics.js:206]: The error message needs to be more helpful to the user.

```suggestion
        throw Error(`${intg}: The destination configuration parameter "${config}" is either unavailable or invalid. Please check your destination settings in the dashboard.`);
```

File: src/core/analytics.js [hunk 3]
```diff
@@ -195,6 +196,17 @@ class Analytics {
     );
   }
 
+  validateConfig(requiredConfigsList, configsFromDashboard, intg) {
+    if(!requiredConfigsList){
+      throw Error(`${intg}: required configs not present`);
```
Comment (saikumarrs) [inline src/core/analytics.js:201]: Same as the below comment.

File: src/integrations/Adroll/browser.js [hunk 4]
```diff
@@ -17,6 +17,7 @@ class Adroll {
     this.analytics = analytics;
     this.advId = config.advId;
     this.pixId = config.pixId;
+    this.requiredConfigsList = requiredConfigsList;
```
Comment (saikumarrs) [inline src/integrations/Adroll/browser.js:20]: Why does it have to be defined in the class? We don't seem to use this.

**Review by saikumarrs — CHANGES_REQUESTED** [review]
What's the need to duplicate the required fields info in the SDKs and also validate again here? For destinations that are created a long time ago?
CC @krishna2020

**Comment by saikumarrs** [comment]
Also, this is not a `chore`. We're adding a new feature to the SDK. So, let's correct the PR title.

File: src/integrations/Adroll/browser.js (hunk 4 above)
Comment (anantjain45823) [reply src/integrations/Adroll/browser.js:20]: Removing

**Pull request closed without merging**